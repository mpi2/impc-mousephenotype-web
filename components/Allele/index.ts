export { default as AlleleMap } from './AlleleMap';
export { default as Crispr } from './Crispr';
export { default as ESCell } from './ESCell';
export { default as IVP } from './IVP';
export { default as Mice } from './Mice';
export { default as QCModal } from './QCModal';
export { default as TVP } from './TVP';
