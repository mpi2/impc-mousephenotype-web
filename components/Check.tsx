import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck, faXmark } from "@fortawesome/free-solid-svg-icons";

const Check = ({ isChecked }) => {
  const styles = isChecked ? {} : { backgroundColor: '#8E8E8E' }
  return (
    <span style={styles}>
      <FontAwesomeIcon icon={isChecked ? faCheck : faXmark} style={{ color: '#000' }} />
    </span>
  )
};

export default Check;
