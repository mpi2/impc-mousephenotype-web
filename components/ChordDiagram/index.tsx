import * as d3 from 'd3';
import { useEffect, useRef } from "react";
import styles from './styles.module.scss';

interface Props  {
  // data: Array<Array<number>>;
  width: number;
  height: number;
  topTerms?: Array<string>;
}

function groupTicks(d, step) {
  var k = (d.endAngle - d.startAngle) / d.value;
  return d3.range(0, d.value, step).map(function (value) {
    return {value: value, angle: value * k + d.startAngle, index: d.index};
  });
}

const ChordDiagram = ({ width, height, topTerms }: Props) => {
  const ref = useRef();
  const outerRadius = Math.min(width, height) * 0.5 - 200;
  const innerRadius = outerRadius - 30;
  // TODO: to be removed after getting payload for landing pages
  const matrix = [
    [
      590,
      268,
      712,
      64,
      45
    ],
    [
      268,
      0,
      26,
      12,
      4
    ],
    [
      712,
      26,
      0,
      11,
      6
    ],
    [
      64,
      12,
      11,
      0,
      0
    ],
    [
      45,
      4,
      6,
      0,
      0
    ]
  ]
  let labels = "[{\"name\":\"cardiovascular system phenotype\",\"geneCount\":1622}, {\"name\":\"vision\\\/eye phenotype\",\"geneCount\":268}, {\"name\":\"growth\\\/size\\\/body region phenotype\",\"geneCount\":712}, {\"name\":\"embryo phenotype\",\"geneCount\":64}, {\"name\":\"muscle phenotype\",\"geneCount\":45}]";
  let labels_json = JSON.parse(labels);

  useEffect(() => {
    const svgElement = d3.select(ref.current);

    function fade(opacity) {
      return function (d, i) {
        const { index } = i;
        // hide other chords on mose over
        svgElement.selectAll("path.chord")
          .filter(chord => chord.source.index !== index && chord.target.index !== index)
          .transition()
          .style("stroke-opacity", opacity)
          .style("fill-opacity", opacity);
      };
    }
    const chord = d3.chord()
      .padAngle(0.05)
      .sortSubgroups(d3.descending);

    const arc = d3.arc().innerRadius(innerRadius).outerRadius(outerRadius);

    const ribbon = d3.ribbon().radius(innerRadius);

    const color = d3.scaleOrdinal()
      .domain(d3.range(4))
      .range([
        "rgb(239, 123, 11)",
        "rgb(9, 120, 161)",
        "rgb(119, 119, 119)",
        "rgb(238, 238, 180)",
        "rgb(36, 139, 75)",
        "rgb(191, 75, 50)",
        "rgb(255, 201, 67)",
        "rgb(191, 151, 50)",
        "rgb(239, 123, 11)",
        "rgb(247, 157, 70)",
        "rgb(247, 181, 117)",
        "rgb(191, 75, 50)",
        "rgb(151, 51, 51)",
        "rgb(144, 195, 212)"
      ]);

    const g = svgElement.append("g")
      .attr("transform", `translate(${width / 2}, ${height / 2})`)
      .datum(chord(matrix));

    const group = g.append("g")
      .attr("class", "groups")
      .selectAll("g")
      .data(chords => chords.groups)
      .enter().append("g")
      .attr("class", "group")
      .on("mouseover", fade(.02))
      .on("mouseout", fade(.80));

    group.append("path")
      .style("fill", d => color(d.index))
      .style("stroke", d => d3.rgb(color(d.index)).darker())
      .attr("d", arc);

    const groupTick = group.selectAll(".group-tick")
      .data(d => groupTicks(d, 1e3))
      .enter().append("g")
      .attr("class", "group-tick")
      .attr("transform", d => `rotate(${d.angle * 180 / Math.PI - 90}) translate(${outerRadius},0)`);

    groupTick.append("line")
      .attr("x2", 6);

    groupTick
      .filter(d => d.value % 5e3 === 0)
      .append("text")
      .attr("x", 8)
      .attr("dy", ".35em")
      .attr("transform", d => d.angle > Math.PI ? "rotate(180) translate(-16)" : null)
      .style("text-anchor", d => d.angle > Math.PI ? "end" : null)
      .text(d => labels_json[d.index].name.replace("phenotype", ""));

    g.append("g")
      .attr("class", "ribbons")
      .selectAll("path")
      .data(chords => chords)
      .enter()
      .append("path")
      .attr("d", ribbon)
      .attr("class", "chord")
      .style("fill", d => color(d.target.index))
      .style("stroke", d =>d3.rgb(color(d.target.index)).darker())
      .style("visibility", d => {
        if (topTerms && topTerms.length > 0) {
          if (topTerms.indexOf(labels_json[d.source.index].name) < 0 && topTerms.indexOf(labels_json[d.target.index].name) < 0) {
            return "hidden";
          } else {
            return "visible";
          }
        } else {
          return "visible";
        }
      })
      .append("title").text(d => {
        return d.source.value + " genes present " + labels_json[d.source.index].name + " and " + labels_json[d.target.index].name + ", " ;
      });

  }, []);

  return (
    <div className={styles.wrapper}>
      <svg
        width={width}
        height={height}
        ref={ref}
      />
    </div>
  )
};

export default ChordDiagram;