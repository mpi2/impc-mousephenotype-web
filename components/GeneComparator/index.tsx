import { useRouter } from "next/router";
import React, { useEffect, useState } from "react";
import { Button } from "react-bootstrap";

const GeneComparatorContext = React.createContext(null);

export const GeneComparatorProvider = ({ children }) => {
  const [genes, setGenes] = useState([]);

  const addGene = (gene) => {
    if (gene && !genes.includes(gene)) {
      const newGenes = [...genes, gene];
      setGenes(newGenes);
      return newGenes;
    }
    return genes;
  };

  const removeGene = (gene) => {
    if (gene && genes.includes(gene)) {
      genes.splice(genes.indexOf(gene));
      setGenes(genes);
    }
  };

  const resetGenes = (genes: string[]) => {
    setGenes(genes);
  };

  return (
    <GeneComparatorContext.Provider
      value={{ genes, addGene, removeGene, resetGenes }}
    >
      {children}
    </GeneComparatorContext.Provider>
  );
};

export const useGeneComparator = () => React.useContext(GeneComparatorContext);

export const GeneComparatorTrigger = ({ current }: { current?: string }) => {
  const { addGene, genes } = useGeneComparator();
  const router = useRouter();
  return (
    <Button
      style={{
        position: "fixed",
        zIndex: 100,
        bottom: genes.length > 0 || current ? 0 : -100,
        right: 100,
        transition: ".3s ease transform",
      }}
      variant="primary"
      onClick={() => {
        if (current) {
          const updatedGenes = addGene(current);
          return router.push(`/data/compareGenes/?genes=${updatedGenes.join("_")}`);
        }
        router.push(`/data/compareGenes/?genes=${genes.join("_")}`);
      }}
    >
      Compare Genes{" "}
      <span
        className="bg-white"
        style={{
          display: "inline-block",
          height: 22,
          padding: "0 7px",
          borderRadius: 11,
          marginLeft: 6,
          fontWeight: "bold",
          color: '#000',
        }}
      >
        {genes.length}
      </span>
    </Button>
  );
};
