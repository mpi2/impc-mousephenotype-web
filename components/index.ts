export { default as Card } from './Card';
export { default as Search } from './Search';
export { default as AlleleSymbol } from './AlleleSymbol';
export { default as DownloadData } from './DownloadData'