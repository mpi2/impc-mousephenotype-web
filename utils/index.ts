import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};
