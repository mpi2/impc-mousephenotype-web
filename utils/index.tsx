import React from "react";
import _ from "lodash";

export const formatBodySystems = (systems: string[] | string = []) => {
  return _.capitalize(
    (typeof systems === "string" ? systems : systems.join(", "))
      .replace(/ phenotype/g, "")
      .replace(/\//g, " / ")
  );
};

export const formatAlleleSymbol = (allele: string) => {
  return allele.slice(0, allele.length - 1).split("<");
};

export const formatPValue = (pValue: number) => {
  const pValueArray = Number.parseFloat(String(pValue))
    .toExponential(2)
    .split("e");
  return (
    <>
      {pValueArray[0]}x10<sup>{pValueArray[1].replace("+", "")}</sup>
    </>
  );
};

export const formatESCellName = (src: string) => {
  const [name, superscript, end] = src.split(/<|>/);
  return (
    <>
      {name}
      <sup>{superscript}</sup>
      {end}
    </>
  );
};

const fetchCache = {};
export { fetchCache };

export const toSentenceCase = (camelCase) => {
  if (camelCase) {
    const result = camelCase.replace(/([A-Z])/g, " $1");
    return result[0].toUpperCase() + result.substring(1).toLowerCase();
  }
  return "";
};
