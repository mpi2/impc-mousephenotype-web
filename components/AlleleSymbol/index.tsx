import { formatAlleleSymbol } from "@/utils";

const AlleleSymbol = ({ symbol, withLabel = true }) => {
  const allele = formatAlleleSymbol(symbol);
  return <span>{!!withLabel ? 'Allele: ' : ''}{allele[0]}<sup>{allele[1]}</sup></span>
}

export default AlleleSymbol;