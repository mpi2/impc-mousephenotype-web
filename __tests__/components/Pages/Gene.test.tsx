import renderer from 'react-test-renderer';
import GenePage from "@/pages/genes/[pid]";
import { QueryClientProvider } from "@tanstack/react-query";
import { createTestQueryClient } from "../../utils";
import mockRouter from "next-router-mock";
import { GeneComparatorProvider } from "@/components/GeneComparator";

jest.mock('next/router', () => jest.requireActual('next-router-mock'));

describe('Gene page', () => {
  it('renders correctly', async () => {
    const client = createTestQueryClient();
    await mockRouter.push('/genes/MGI:1336993');
    const tree = renderer.create(
      <QueryClientProvider client={client}>
        <GeneComparatorProvider>
          <GenePage />
        </GeneComparatorProvider>
      </QueryClientProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});