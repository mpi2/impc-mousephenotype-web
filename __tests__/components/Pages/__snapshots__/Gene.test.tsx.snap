// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gene page renders correctly 1`] = `
<div>
  <button
    class="btn btn-primary"
    style="position: fixed; z-index: 100; bottom: -100px; right: 100px; transition: .3s ease transform;"
    type="button"
  >
    Compare Genes
     
    <span
      class="bg-white"
      style="display: inline-block; height: 22px; padding: 0px 7px; border-radius: 11px; margin-left: 6px; font-weight: bold; color: rgb(0, 0, 0);"
    >
      0
    </span>
  </button>
  <div
    class="banner"
  >
    <div
      class="pb-4 pt-5 container container"
    >
      <div
        class="col-12 col-md-8 ps-4 pe-4"
      >
        <div
          class="tabs"
        >
          <a
            class="tab__active"
            href="/search"
          >
            Genes
          </a>
          <a
            class="tab"
            href="/search?type=phenotype"
          >
            Phenotypes
          </a>
          <a
            class="tab"
            href="https://dev.mousephenotype.org/?s="
          >
            Help, news, blog
          </a>
        </div>
        <div
          class="inputCont"
        >
          <input
            class="input"
            placeholder="Search All 7824 Knockout Data..."
            title="main search box"
            type="text"
            value=""
          />
          <button
            aria-describedby="svg-inline--fa-title-search-icon"
            class="searchBtn"
          >
            <svg
              aria-labelledby="svg-inline--fa-title-search-icon"
              class="svg-inline--fa fa-magnifying-glass "
              data-icon="magnifying-glass"
              data-prefix="fas"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title
                id="svg-inline--fa-title-search-icon"
              >
                Search button
              </title>
              <path
                d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                fill="currentColor"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="page container"
  >
    <div
      class="card"
    >
      <div
        class="subheadingCont"
      >
        <div
          class="subheading"
        >
          <span
            class="subheadingSection"
          >
            Gene
          </span>
          <span
            class="subheadingSection"
          />
        </div>
      </div>
      <br />
      <p
        class="grey"
      >
        Loading...
      </p>
    </div>
  </div>
</div>
`;
